source/brailleViewer/__init__.py

```diff
@@ -0,0 +1,112 @@
+connected physical device.
+
+### Life-cycle
+- Constructing / showing the BrailleViewer
+	- On startup via L{core.doStartupDialogs}
+	- Via NVDA (tools) menu via L{Mainframe.onToggleSpeechViewerCommand}
+- Hiding / destroying the BrailleViewer
+	- On exit of NVDA.
+	- Via NVDA (tools) menu via L{Mainframe.onToggleSpeechViewerCommand}
+	- When the Window receives a close event. This means the GUI must be able to call-back to clean up
+	BrailleHandler and the NVDA tools menu. This callback happens via the L{postBrailleViewerToolToggledAction}
+
+### Number of cells shown
+The default (40) is set in L{createBrailleViewerTool}.
+
+### Routing
+Currently not supported.
+In order to support routing the user must be able to click on the cells. This means that the BrailleViewer
+window gains focus, and the braille values are changed. To avoid this would require substantial changes to
+brailleHandler.
+
+### Scrolling
+Scrolling is supported by binding a gesture to the braille_scroll_forward and braille scroll_back commands.
+For the same reason that Routing is not supported, scrolling via button clicks on the braille viewer window
+is not supported.
+
+"""
+
+# global braille viewer driver:
+_brailleGui: Optional[BrailleViewerFrame] = None
+
+# Extension points action:
+# Triggered every time the Braille Viewer is created / shown or hidden / destroyed.
+# Callback definition: Callable(created: bool) -> None
+#   created - True for created/shown, False for hidden/destructed.
+postBrailleViewerToolToggledAction = extensionPoints.Action()
+# Devices with 40 cells are quite common.
+DEFAULT_NUM_CELLS = 40
+
+
+def isBrailleViewerActive() -> bool:
+	return bool(_brailleGui)
+
+
+def update(cells: List[int], rawText: str):
+	if _brailleGui:
+		_brailleGui.updateBrailleDisplayed(cells, rawText)
+
+
+def _destroyGUI():
+	global _brailleGui
+	d: Optional[BrailleViewerFrame] = _brailleGui
+	_brailleGui = None
+	if d and not d.isDestroyed:
+		d.Destroy()
+
+
+def destroyBrailleViewer():
+	_destroyGUI()
+	postBrailleViewerToolToggledAction.notify(created=False)
+
+
+def _onGuiDestroyed():
+	""" Used as a callback from L{BrailleViewerFrame}, lets us know that the GUI initiated a destruction.
+	"""
+	global _brailleGui
+	if _brailleGui:
+		# Destruction wasn't initiated from L{_destroyGUI} ie not through direct user action.
+		# It's likely that the window received a shutdown event that could not be skipped.
+		# Continue to destroy the braille viewer.
+		destroyBrailleViewer()
+
+
+def createBrailleViewerTool():
+	if not gui.mainFrame:
+		raise RuntimeError("Can not initialise the BrailleViewerGui: gui.mainFrame not yet initialised")
+
+	import braille  # imported late to avoid a circular import.
+	if not braille.handler:
+		raise RuntimeError("Can not initialise the BrailleViewerGui: braille.handler not yet initialised")
+
+	numCells = DEFAULT_NUM_CELLS
+	if braille.handler.displaySize:
+		numCells = braille.handler.displaySize

```

Comment:
It worries me slightly that this sets numCells according to braille.handler.displaySize, which itself also checks the braille viewer. I think it will work as expected, though.

Reply:
Yes, this can now be simplified.